### PR TITLE
api: add rule middleware

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #5839

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/api/rule.go [hunk 1]
```diff
@@ -42,6 +43,42 @@ func newRulesHandler(svr *server.Server, rd *render.Render) *ruleHandler {
...
+
+func newRuleMiddleware(s *server.Server, rd *render.Render) ruleMiddleware {
+	return ruleMiddleware{
+		s:       s,
+		rd:      rd,
+		Handler: s.GetHandler(),
+	}
+}
+
+func (m ruleMiddleware) Middleware(h http.Handler) http.Handler {
```
Comment (lhy1024) [inline server/api/rule.go:60]: Will we support it in scheduling server?

Comment (CabinfeverB) [reply server/api/rule.go:60]: scheduling server uses `gin` and does implement this in `Handler`. So i think we don't need to use middleware in scheduling server now.

File: server/api/rule.go [hunk 2]
```diff
@@ -42,6 +43,42 @@ func newRulesHandler(svr *server.Server, rd *render.Render) *ruleHandler {
 	}
 }
 
+type ruleMiddleware struct {
+	s  *server.Server
+	rd *render.Render
+	*server.Handler
+}
+
+func newRuleMiddleware(s *server.Server, rd *render.Render) ruleMiddleware {
```
Comment (rleungx) [inline server/api/rule.go:52]: Shall we move it to middleware.go?

Comment (CabinfeverB) [reply server/api/rule.go:52]: Because it is only used in rulesHander, I put it in rules.go.

Comment (rleungx) [reply server/api/rule.go:52]: Both are ok to me.

**Review by rleungx — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**